File: apps/web/src/types/studentOwnExercise.ts
import type { DifficultyLevel, Language, MathProblemAnalysis } from './problem';

export interface StudentOwnExercise {
  id: string;
  userId: string;
  problemText: string;
  analysis: MathProblemAnalysis;
  language: Language;
  difficultyLevel: DifficultyLevel;
  createdAt: string;
  updatedAt: string;
}

export interface StudentOwnExerciseCreate {
  problemText: string;
  analysis: MathProblemAnalysis;
  language: Language;
}

export interface StudentOwnExerciseUpdate {
  problemText?: string | null;
  analysis?: MathProblemAnalysis | null;
  language?: Language | null;
}

File: apps/api/app/models/student_own_exercise.py
"""Student-created own exercises model."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from sqlalchemy import DateTime, ForeignKey, JSON, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base, utcnow
from .user import User


class StudentOwnExercise(Base):
    """Student-created math word problem exercise."""

    __tablename__ = "student_own_exercises"

    id: Mapped[str] = mapped_column(String(36), primary_key=True)
    user_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    problem_text: Mapped[str] = mapped_column(Text, nullable=False)
    analysis: Mapped[dict[str, Any]] = mapped_column(JSON, nullable=False)
    language: Mapped[str] = mapped_column(String(5), nullable=False, default="en")
    difficulty_level: Mapped[str] = mapped_column(String(10), nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=utcnow, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=utcnow, onupdate=utcnow, nullable=False
    )

    # Relationship
    user: Mapped[User] = relationship(back_populates="own_exercises")
